rasterio/_crs.pyx

```diff
@@ -50,17 +65,24 @@ class _CRS(UserDict):
         cdef int retval
 
         try:
-            if not self and not other:
-                return True
+            if not self or not other:
+                return not self and not other
 
             # use dictionary equality rules first
-            elif UserDict(self.data) == UserDict(other):
-                return True
-
-            elif not self or not other:
-                return False
+            elif isinstance(other, dict):
+                warnings.warn(
+                    "Comparison to dict will be removed in 1.0",
+                    RasterioDeprecationWarning
+                )
+                if UserDict(self.data) == UserDict(other):
+                    return True
 
             else:
+                if isinstance(other, str):
+                    warnings.warn(
+                        "Comparison to str will be removed in 1.0",
+                        RasterioDeprecationWarning
+                    )
```

Comment:
@sgillies not sure to understand why we depreciate the dict equality

Reply:
Because `dict == CRS(...)` won't work without extra magic, I think it's simpler to require `CRS()` on each side of the comparison.